### PR TITLE
Fixed issue #18146: Unable to use "other" as subquestion code

### PR DESCRIPTION
**Pull request by Shnoulle**
Dev: reset other to N when save by rules
Dev: in 3.X : done in database, getPost is empty when not submitted
Dev: inverse must be done too when try to update to single choice with other

### CODE REVIEW

**Comment by Shnoulle** [comment]
Maybe set to null when not applicable ?

**Comment by Shnoulle** [comment]
Usage of questiontype : set it as draft temporary

File: application/models/Question.php [hunk 1]
```diff
@@ -144,6 +144,12 @@ public function rules()
             array('qid,sid,gid,parent_qid', 'numerical', 'integerOnly' => true),
             array('qid', 'unique','message' => sprintf(gT("Question id (qid) : '%s' is already in use."), $this->qid)),// Still needed ?
             array('other', 'in', 'range' => array('Y', 'N'), 'allowEmpty' => true),
+            array('other', 'filter', 'filter' => function ($value) {
```
Comment (gabrieljenik) [inline application/models/Question.php:147]: I believe this shall have a bigger sign. Like a separate block of lines inside the rules() method, as for highlighting this is a filter. (which is not frequently used)

File: application/models/Question.php [hunk 2]
```diff
@@ -176,10 +182,12 @@ public function rules()
                 $aRules[] = array('title', 'match', 'pattern' => '/comment$/', 'not' => true, 'message' => gT("'comment' suffix can not be used with multiple choice with comments."));
             }
         } else {
-            // Disallow other if sub question have 'other' for title
-            $oSubquestionOther = Question::model()->find("parent_qid=:parent_qid and LOWER(title)='other'", array("parent_qid" => $this->qid));
-            if ($oSubquestionOther) {
-                $aRules[] = array('other', 'compare', 'compareValue' => 'Y', 'operator' => '!=', 'message' => sprintf(gT("'%s' can not be used if the 'Other' option for this question is activated."), 'other'), 'except' => 'archiveimport');
+            if ($this->getAsSubquestions()) {
+                // Disallow other if sub question have 'other' for title
+                $oSubquestionOther = Question::model()->find("parent_qid=:parent_qid and LOWER(title)='other'", array("parent_qid" => $this->qid));
```
Comment (gabrieljenik) [inline application/models/Question.php:187]: Maybe we can wrap this into a method?

File: application/models/Question.php [hunk 3]
```diff
@@ -176,10 +182,12 @@ public function rules()
...
+                if ($oSubquestionOther) {
+                    $aRules[] = array('other', 'compare', 'compareValue' => 'Y', 'operator' => '!=', 'message' => sprintf(gT("'%s' can not be used if the 'Other' option for this question is activated."), 'other'), 'except' => 'archiveimport');
```
Comment (gabrieljenik) [inline application/models/Question.php:189]: Can you remind me why the archiveImport is excepted?
Maybe add it as a comment?

**Review by gabrieljenik — COMMENTED** [review]
Left a few comments! Nice PR!

File: application/models/Question.php (hunk 3 above)
Comment (Shnoulle) [reply application/models/Question.php:189]: I think it's for LSA

File: application/models/Question.php (hunk 1 above)
Comment (Shnoulle) [reply application/models/Question.php:147]: Just separate in same function ? Can be an idea :)

Keep it at anonymous function or not ? Here , i think it's cleaner enough at anonymous function (and usage only for other)

File: application/models/Question.php (hunk 2 above)
Comment (Shnoulle) [reply application/models/Question.php:187]: you mean something like `$oSubquestionOther = $this->getAsOtherSubquestion()` ?

Can be better, right.

File: application/models/Question.php (hunk 3 above)
Comment (Shnoulle) [reply application/models/Question.php:189]: > I think it's for LSA

but unsure … let's grep

File: application/models/Question.php (hunk 1 above)
Comment (gabrieljenik) [reply application/models/Question.php:147]: >> Just separate in same function ? Can be an idea :)

Yes, just as a separate block of lines inside the function

>> Keep it at anonymous function or not ? Here , i think it's cleaner enough at anonymous function (and usage only for other)

Maybe a setOther method?

File: application/models/Question.php (hunk 2 above)
Comment (gabrieljenik) [reply application/models/Question.php:187]: why are there methods names like `getAsxxxx`?
What do getAsxxx means?

File: application/models/Question.php (hunk 3 above)
Comment (gabrieljenik) [reply application/models/Question.php:189]: I would add a comment about it.

File: application/models/Question.php (hunk 1 above)
Comment (Shnoulle) [reply application/models/Question.php:147]: > Maybe a `setOther()` method?

Think it less clear what it's done if not in existing rule …

File: application/models/Question.php (hunk 2 above)
Comment (gabrieljenik) [reply application/models/Question.php:187]: `hasSubquestionsOther` ?
Do we need the `get` prefix?

File: application/models/Question.php (hunk 3 above)
Comment (Shnoulle) [reply application/models/Question.php:189]: And here : i need to remove archiveimport allowed : it broke DB :)

File: application/models/Question.php (hunk 2 above)
Comment (Shnoulle) [reply application/models/Question.php:187]: `$question->hasubquestionother` == `$question->getHasOtherSubquestion()` no ?

File: application/models/Question.php (hunk 1 above)
Comment (gabrieljenik) [reply application/models/Question.php:147]: > > Maybe a `setOther()` method?
> 
> Think it less clear what it's done if not in existing rule …

I meant: call setOther from the rules

Comment (Shnoulle) [reply application/models/Question.php]: ? We have 'filter' rules, then adding a function like `addRulesFilterOther` ? don't like.

I prefer to create
`addRulesSubQuestiion` and `haveRulesPrimaryQuestion` in this case.

File: application/models/Question.php (hunk 2 above)
Comment (Shnoulle) [reply application/models/Question.php]: https://github.com/LimeSurvey/LimeSurvey/pull/2443/files#diff-75d3428e7aa3f16b9dc93ef54dfa4d0556c04431cb53510749ded531d549a148R1488

**Comment by Shnoulle** [comment]
@c-schmitz why you remove bug fix ? There are a clearly a bug in 5.X , this bug didn't exist in 3.X

File: application/models/Question.php (hunk 1 above)
Comment (Shnoulle) [reply application/models/Question.php]: Check please :)

Comment (gabrieljenik) [reply application/models/Question.php]: I mean you could have like this:
```
function setOther($value) {
 if ($this->getAllowOther()) {
          return $value;
      }
      return 'N';
  }
}
```

and then from within rules
```
array('other', 'filter', 'filter' => function ($value) { return $this->SetOther($value); }
```

File: application/models/Question.php (hunk 2 above)
Comment (gabrieljenik) [reply application/models/Question.php]: You could have like this 
```
function getSubquestionOther() {
   return Question::model()->find("parent_qid=:parent_qid and LOWER(title)='other'", array("parent_qid" => $this->qid));
}
```

then use `getSubquestionOther()` in the code.

File: application/models/Question.php (hunk 1 above)
Comment (Shnoulle) [reply application/models/Question.php]: > Not sure is the best, but thats what I meant originally.

Compare current filter on other and filter on theme : and tell me the simple one … I think anonymous function is clearer here.

Maybe @olleharstedt or @c-schmitz advice ?

Comment (gabrieljenik) [reply application/models/Question.php]: Just keep it as it is.
No big deal!

**Review by gabrieljenik — APPROVED** [review]

File: application/models/Question.php (hunk 1 above)
Comment (Shnoulle) [reply application/models/Question.php]: vote ;) 

I'm OK to create a separate function, hard to know what is the better solution. 

Create 2 function : addSubquestionsRules and addPrimaryQuestionRules seems better to do before.

**Comment by gabrieljenik** [comment]
https://bugs.limesurvey.org/view.php?id=18146#c70337

File: application/models/Question.php (hunk 1 above)
Comment (olleharstedt) [reply application/models/Question.php]: You can also write a function that produces functions. :)

Comment (Shnoulle) [reply application/models/Question.php]: The question is : did we really need to construct some thing more complex than 
```

        $aRules[] = array('other', 'filter', 'filter' => function ($value) {
            if ($this->getAllowOther()) {
                return $value;
            }
            return 'N';
        });
```

When we don't have a better separation of rules for subquestions and questions for example.

**Comment by gabrieljenik** [comment]
Tested OK.  Haven't done regression testing.
Still considering the strange reproduction steps needed and the code updated, as to avoid regressions, I would send it to DEV as to have the code be sinked and lived for a while.

**Comment by Shnoulle** [comment]
>  the strange reproduction steps 

? Just set other to a question, never resetted to no when updating the question.

Mazi have the issue in real survey … and can not fix without hacking DB … 

It's a big issue, really a big issue …
And we don't have this issue in 3.X …

**Pull request merged**